### ISSUE TITLE
Strict mode for CBOR

### ISSUE THREAD

**Issue by timmc** [Kotlin/kotlinx.serialization#2662]
**What is your use-case and why do you need this feature?**

My app needs to be able to sign and verify serialized data, which requires that the parse be unambiguous. The main problem is duplicate map keys. If there are repeated keys, then two recipients may disagree on what the signed data says, which is obviously a problem. :-)

CBOR's spec describes a "strict mode" that forbids repeated map keys and other dangers to canonicalization: https://datatracker.ietf.org/doc/html/rfc7049#section-3.10

**Describe the solution you'd like**

1. Generally support rejection of repeated map keys, as an optional feature. (I'd love for this to be enabled by default, but I understand that there are competing needs here.)
2. Expose this as part of a new strict-mode feature in the `Cbor` implementation.

Note: https://github.com/Kotlin/kotlinx.serialization/issues/1990 covers rejection of duplicate keys, but it wasn't clear if that was meant to be specific to JSON.

**Comment by timmc** [comment]
(Updated with more info from the spec.)

**Comment by sandwwraith** [comment]
Yes, it sounds very reasonable. I do not mind adding it at some point in the future or via contribution.

**Comment by timmc** [comment]
Great, I may try to work up a PR. Where would you recommend getting started? I found [`insertKeyValuePair` in CollectionSerializers](https://github.com/Kotlin/kotlinx.serialization/blob/51cb8e8e556983fc83a565d5f04bb089363453e0/core/commonMain/src/kotlinx/serialization/internal/CollectionSerializers.kt#L265)—but it didn't seem that the tests actually exercise that code path, and I can't find anything that calls `insertKeyValuePair`.

**Comment by pdvrieze** [comment]
@timmc What happens from a format perspective is that it recognizes that this is a collection serializer. Then it will call the update decoder (at some point it was part of DeserializerStrategy, but it is no longer) with the previous list value.That update will create/update the collection with the new value (or not). It would be possible to create a new version of this function that takes a strict argument (with the default implementation of the previous one passing false for strictness).

**Comment by timmc** [comment]
Would you be able to point me to a code location? I'm not finding useful references to `update`, `CollectionSerializer`, etc. I just discovered that `LinkedHashMapSerializer` extends `LinkedHashMap`, but I'm having trouble finding the code that actually puts key-value pairs into a map when deserializing map data.

**Comment by timmc** [comment]
OK, found it -- `builder[key] = value` in `MapLikeSerializer`. Starting a branch here: https://github.com/Kotlin/kotlinx.serialization/compare/dev...timmc:timmc/cbor-strict

**Comment by timmc** [comment]
I've found it straightforward to reject duplicate keys. Making it _configurable_ would be a larger undertaking—I can either plumb it through via the serializers or via the descriptors, but there's not really a clean way to do it since there's no config object being passed along. Instead, I would have to add `allowDuplicateKeys: Boolean` to a large number of method and constructor signatures, and I still haven't found the end of that chain.

Would you accept a PR that just forbids duplicate keys in maps and objects, with an option to later add a configuration option that permits duplicate keys?

**Comment by sandwwraith** [comment]
@timmc I don't think it is a good idea to change existing behavior so that people would start receiving exceptions on a code that worked before, especially without an option to bring old behavior back. So it's better to have it. Flags are already stored in `Cbor` object, which is available in every `CborReader`. So you need to pass this object further down to `CborDecoder`, and that's likely it.

**Comment by pdvrieze** [comment]
@timmc The code involved is:

https://github.com/Kotlin/kotlinx.serialization/blob/51cb8e8e556983fc83a565d5f04bb089363453e0/core/commonMain/src/kotlinx/serialization/internal/CollectionSerializers.kt#L99-L113

and

https://github.com/Kotlin/kotlinx.serialization/blob/51cb8e8e556983fc83a565d5f04bb089363453e0/core/commonMain/src/kotlinx/serialization/internal/CollectionSerializers.kt#L26-L51

It is important to note that for some formats it is valid to read list/map items one by one (for example in Protobuf):
https://github.com/Kotlin/kotlinx.serialization/blob/51cb8e8e556983fc83a565d5f04bb089363453e0/formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt#L210-L218

This later case requires special handling of the MapSerializer. So you have two separate cases to support this generically (reading multiple items at a time, directly in MapSerializer; and repeatedly reading sublists).

**Comment by sandwwraith** [comment]
I'm not sure it is a good idea to add changes to `CollectionSerializer` since it is used by all formats. You also need to check whether regular objects have duplicate keys. It seems you have to track keys manually, with some kind of `Set`.

**Comment by pdvrieze** [comment]
@sandwwraith Indeed modifying CollectionSerializer would be a major undertaking with various compatibility concerns/challenges.

**Comment by timmc** [comment]
I don't think `CborMapReader` has enough information to build and check a list of keys. By the time `decodeElementIndex` has been called (in the parent class `CborListReader`), `MapLikeSerializer.readElement` has _already_ called `decodeSerializableElement` to read the key.

Maybe it would make sense for `MapLikeSerializer` to call a new `CompositeDecoder.visitedKey` method that accepts a key and then has the option of throwing an exception?

**Comment by pdvrieze** [comment]
> Maybe it would make sense for `MapLikeSerializer` to call a new `CompositeDecoder.visitedKey` method that accepts a key and then has the option of throwing an exception?

I like this solution (even though some bits of the name/semantics could be "improved"). The function can have a default implementation that results in the current behaviour. Note that you should probably also think about the set serializer (sets also don't allow duplicate values/only keep the last one).

**Comment by timmc** [comment]
Yeah, it seems to work out OK -- I have https://github.com/Kotlin/kotlinx.serialization/pull/2681 which still needs some work but should be on the right path.

**Comment by timmc** [comment]
I'm coming back to this work and before I try to get my PR updated against recent changes, I'm wondering if there's still  any interest in this issue? There are some significant merge conflicts and I don't want to work through them if the changes would be unwelcome.

I also noticed that there is now a COSE compliance mode, but the COSE spec indicates that parsing duplicate keys is illegal:

> Applications MUST NOT parse and process messages with the same label used twice as a key in a single map.

(from https://www.rfc-editor.org/rfc/rfc8152.txt section 14 "CBOR Encoder Restrictions")

Did this restriction get implemented somewhere after all, or is it still missing?
